### DISCUSSION TITLE
RememberSession True but not for 1 ?

### DISCUSSION THREAD

**Discussion by JoePalffy** [sumatrapdfreader/sumatrapdf#2862] [Ideas]
So i have RememberSession set as true and it works if there are more than 1 files opened, but if I only have 1 file open then it will not resume, can that be fixed please ? or am I missing a setting ?

Oh and does SumatraPDF not have thumbnail support ? the "home" page only shows 10 tiny filetype icons instead of thumbnails and no filenames, but they show up in file explorer, this makes the home page next to useless, could we at least have file names/titles show up?

I would also love it if there were more zoom steps (aka finer control), right now the steps between zoom levels are way too big, twice as many steps would already be a massive improvement.

Also when I right click an image, can we have an option to save as image ? right now there is a Copy Image, but that's useless to me.

I should mention that I am using the latest version (although when i got to Help>About it doesn't tell me the version number for some reason, I've never seen a program that fails to show that) and that I'm using the portable version.

(should I have started a new discussion for each of these?)

**Comment by GitHubRulesOK** [comment]
1) One file tab on closure behavior changed a few times recently so check current pre-release for behavior as should be different to say 3.2 ? there is already a formal open issue 
  2) Long time SumatraPDF home page has always had 10 PNG thumbnails of *generally* upper cover page (depends on first page ratio) visible from the cache they can  be replaced "sometimes" but the seemingly random numbering can make personal substitution disappear when renaming the source file. again there is already a related still open issue or two :-)
![image](https://user-images.githubusercontent.com/12754913/184651461-5643c679-9db7-4371-a7ee-7340874e6ae9.png)

  3) Zoom steps can be customized as either ZoomIncrement = #% (perhaps too variable as not pure linear integer values) or as current by adding/editing step sizes ZoomLevels = 8.33 10 20 30 40 50 .... (min 8.3333 max 6400) however if its not a listed zoom factor the only way to know is use CTRL+Y to see value so again there are related requests to show in UI/toolbar
![image](https://user-images.githubusercontent.com/12754913/184660652-884a9a0e-8420-40bd-9f5d-bd3c01ff9a87.png)

  4) Several open requests but save page as image is not that simple for all cases, if there is an embedded image the clipboard will take scale into account (try zoom copy and paste screen render in paint, but try to export a pdf image with alpha mask is two (or more) bitmap images overlaid so can often be problematic. simplest for me is export page number to a command line tool like a graphics editor or pdf image extractor since the image may not even be stored in the page but only a cross reference. 

Help about should show version data but if compiled by you that may be always a fixed version such as 3.5
![image](https://user-images.githubusercontent.com/12754913/184662372-099f4dd0-e552-4f31-b100-e0abcd769bc9.png)

**Reply by JoePalffy** [reply]
1. I did not compile it myself, I just used the straight up windows installer.
<img width="297" alt="2022-08-18 13_53_39-About SumatraPDF" src="https://user-images.githubusercontent.com/14361452/185378433-48aa08c3-c6e1-49bb-90bf-eb4ac0976148.png">

2. I'm not sure I understood that, but can't you implement something more simple like what Windows File Explorer does ? or is it a caching issue ? 
Anyway the home page is not something I used in the past, I just thought I'd mention it since it's useless without thumbnails or filenames.

3. Thanks, I just added a bunch more steps, I should have looked more carefully.

4. I wasn't looking for a Save page as Image option, that indeed sounds complicated, I just want to save the embedded image as an image, would that be easier ? (this is from CBZ files, which are full of just images already)
Can you get some inspiration from how STDU Viewer does it ? since that does it really well, it can even export a whole page of text as an image. Unfortunately it hasn't been updated since 2015 and therefor has no WebP support otherwise I would stick to that.

5. Since I use SumatraPDF as a CBZ manga reader it would be really nice if there was some way to open the next chapter more easily. Right now I have to repackage manga with individual chapters into groups of chapters or volumes to avoid opening and closing files 20 times a day. This is only an issue for a few like me, who like reading manga in a continuous scrolling fashion which only PDF reader based manga/comic readers can do (basically only SumatraPDF and STDU Viewer) and mobile apps.

**Reply by GitHubRulesOK** [reply]
Ahh dark on dark that's why you cant see "doesn't tell me the version number for some reason," details
 note there are no theme settings so setting background to black is not a good move for the UI
 (if you must tone down the colors try to use less drastic colour shifts) it would also explain 

2) an ineffective dark on dark home page
3) ok
4) if it is only cbz (not cbr) you could use windows Tar command via filename on a shortcut
 so best to add a cmd file to change as required. Not easy for a novice but here are untested hints, set up a test unCBZ.cmd that at its core should be

Tar -m -xf "%~dpn1.cbz" "%~2"

here is how it should finally work where %1 will be the cbz file and %2 will be the internal tar get

![image](https://user-images.githubusercontent.com/12754913/185386021-b4e7bd19-88e2-4ba4-8e0e-43497bcbf3aa.png)

the problems I forsee are where the file should function and extract to thus safest is first line before tar add `cd /d "%~dp1"` thus move to the folder where the %1 is located to see how that may work start test a small file to extract all without the single page names    

cd /d "%~dp1"
Tar -m -xf "%~dpn1.cbz"
pause

here in my test case drag and drop the cbz on unCBZ.cmd (the two lines above) unpacks all the images (already ran that once here)

![image](https://user-images.githubusercontent.com/12754913/185387787-6676a9b5-4335-4235-96bb-94dcc042c9af.png)


At this point you have a basic unpacker that does not need SumatraPDF to drive it and it can be added into SumatraPDF as an unpack all CBZ images, in the settings file
```
ExternalViewers [
	[
		CommandLine = "C:\Users\WDAGUtilityAccount\Desktop\New folder\unCBZ.cmd" "%1"
		Name = &Unpack all CBZ images
	]
]
```

![image](https://user-images.githubusercontent.com/12754913/185391683-25598caf-0e10-4e2b-84f1-8ec453bf4655.png)

So the much bigger remaining challenge is how to convert page number `%p` into the %2 required to extract one image rather than all

as a first step to understand the deeper problems you can improve the variety of files by change external command to
```
ExternalViewers [
	[
		CommandLine = "C:\Users\WDAGUtilityAccount\Desktop\New folder\unCBZ.cmd" "%1"
		Name = &Unpack all CBZ images / unZip all contents
		Filter = *.cbz;*.fbz;*.zip
	]
]
```
and the cmd file itself to
```
cd /d "%~dp1"
if exist "%~dpn1.cbz" Tar -m -xf "%~dpn1.cbz"
if exist "%~dpn1.fbz" Tar -m -xf "%~dpn1.fbz"
if exist "%~dpn1.zip" Tar -m -xf "%~dpn1.zip"
pause
```
there are other types that could be unpacked by similar method but the problem is knowing how those hundreds of file types were packed so here a zip full of images was packed in a subdir called "images" and that's how they will be expanded
![image](https://user-images.githubusercontent.com/12754913/185398192-55b6f01a-758a-48ad-a57d-40a12b800401.png)

***5) checkout the "next file" commands (CTRL+Shift+RightArrow)***

**Reply by JoePalffy** [reply]
thanks for the reply and sorry for my late reply.

I don't remember changing any colors, I know I wanted the background color while reading manga to be black and the MainWindowBackground setting was indeed set to black, I changed it to light gray (D5D5D5) and now i can see the version number, but the background color behind/next to the manga stayed black, but when I open a PDF then it changes.

So the MainWindowBackground setting applies to PDF files and the UI but not CBZ files, weird. Although I guess that's wrong, it doesn't apply to the UI since the tab and toolbar colors don't change, it's just that the About "page" opens as a document so it gets affected, that should probably not happen or at least it should ignore that color setting and so should the home page.

I looked in the documentation and it says "background color of the non-document windows, traditionally yellow"... isn't that incorrect ? I mean it only applies to document backgrounds (like PDFs) not non-documents like CBZ files, so it's the opposite.

.2 weirdly now the Home page shows proper thumbnails, I was only getting tiny app icons before. Although those 10 thumbnails only take up the upper half of the screen and they are cropped, File Explorer and others usually show the first page in full. 
Could we at least get bigger thumbnails ? like 4 times bigger or is the size of the thumbnails related to my windows scaling options and the size/resolution of my screen?

.5 Oh? Cool, Is there anywhere where I can see what other keyboard shortcuts there are ? and maybe change them ? I'd prefer to use Shift+N for next file (like on Youtube) (and Shift+P for previous) although "<" and ">" are even easier and don't do anything right now.

I saw in the settings file that there is a Shortcuts [ ] line and I read the documentation too where it says to add "Cmd =" and "Key =" ,and I think I know what I should add in the "Key = " field, but I have no idea what goes in the "Cmd =" line or what would be needed for Previous.

But this is a gamechanger, it turns SumatraPDF from a PDF reader that can also read CBZ files into a proper manga reader 😄 it's a shame nobody else knows about this cool feature.
Please consider listing these shortcuts in the Go To dropdown toolbar and if all the shortcuts could be pressed with just one hand that would be great (specially for people with just one hand).

Is there a shortcut to switch between the open tabs ?

And in the View menu it says that the shortcut to Fullscreen is Ctrl+Shift+L but just pressing the F key works too and is so much easier, although I might have set that myself years ago? (i did for other apps) If I did I have no idea how, there is nothing about it in the settings file.

**Reply by GitHubRulesOK** [reply]
black surround is fixed for F5 presentation and image document pages including in zip/comic/manga so not changeable
but for F11 in other formats the surround can be changed using e.g. `GradientColors = #0000FF` up to 3 values allowed to form full document graduation but one alone is suited to all pages.
The about screen etc. is treated as a non document window thus generally follows main setting accordingly. Sidebar, title tabs etc. are windows system colours and thus worked well with XP/7 but MS moved the goalpost with Metro/8/9/10/11, WindowsBlinds will adapt those.

Thumbnail code was written ages ago when 10 half pages often could not fill a small vga screen, time moves on and code still works as it did then. The 10 is based on numeric keys 0-9 so adding more would require extensive changes, agreed bigger may suite those with 16K pixel screens but not those like me still using a netbook where only 4 are visible (that's odd it used to be 8, hmm depends if tabs are on then its 8)

The web guide for commands is interactive https://www.sumatrapdfreader.org/docs/commands start typing the word on right and the options should reduce to the one you need on left

For flip between tabs it should be alt numbers and they used to be draggable sideways (but current pre-release is a bit unstable for both annotations and tabs)

F is usually actioned as Full screen toggle but there are many other legacy ones such that it can be confusing in that and similar cases e.g. only recently C now shown as Continuous toggle and M is not shown for Measure. but I often set Shifted M for toggle Manga mode (for use in R2L PDFs)

**Reply by JoePalffy** [reply]
No, I don't want more than 10 thumbnails, I just wanted them bigger so I can make them out better. 

OK, i added the 2 shortcuts code but I must be missing something (like a separator maybe) since only the first keyboard shortcut works, but if I delete the first one then the 2nd starts working. Then I experimented and figured it out, I needed to add brackets around each shortcut.
Here's what I used:
`Shortcuts [
Cmd = CmdOpenNextFileInFolder
Key = .
]
[
Cmd = CmdOpenPrevFileInFolder
Key = ,
]`
But this was not explained at all on that page, could you fix that when you have time ?
Also the tab switching shortcode (Alt+1, +2 ...) is not listed at all.

**Reply by GitHubRulesOK** [reply]
commands should be per  https://www.sumatrapdfreader.org/docs/Customizing-keyboard-shortcuts [An example of customization:]
```
Shortcuts [
	[
		Cmd = CmdOpenNextFileInFolder
		Key = .
	]
	[
		Cmd = CmdOpenPrevFileInFolder
		Key = ,
	]
]
```
well only firefox could show https://www.sumatrapdfreader.org/docs/Keyboard-shortcuts#search=Tabs  (or SumatraPDF https://www.sumatrapdfreader.org/docs/Keyboard-shortcuts.html then / Tabs but pasting that in file open and its very slow) perhaps save as local
![image](https://user-images.githubusercontent.com/12754913/190279950-7e892446-c93b-4223-82e1-b6458ae3fc65.png)

```
Alt + 1 - go to tab 1. Use Alt + 2, etc. up to Alt + 8
Alt + 9 - go to last tab
```

**Reply by JoePalffy** [reply]
Oh, so the explanations were in a separate document, thanks.
